wkcuber/api/Layer.py

```diff
@@ -293,6 +276,85 @@ def downsample(
     def setup_mag(self, mag: Union[str, Mag]) -> None:
         pass
 
+    def upsample(
+        self,
+        from_mag: Mag,
+        min_mag: Optional[Mag],
+        compress: bool,
+        anisotropic: Optional[bool] = None,
+        scale: Optional[Tuple[float, float, float]] = None,
+        buffer_edge_len: Optional[int] = None,
+        args: Optional[Namespace] = None,
+    ) -> None:
+        assert (
+            from_mag.to_layer_name() in self.mags.keys()
+        ), f"Failed to downsample data. The from_mag ({from_mag}) does not exist."
+
+        if min_mag is None:
+            min_mag = Mag(1)
+
+        if anisotropic and scale is None:
+            if min_mag is None:
+                scale = self.dataset.properties.scale
+            else:
+                scale = calculate_virtual_scale_for_target_mag(min_mag)
+
+        prev_mag = from_mag
+        target_mag = get_previous_mag(prev_mag, scale)
+
+        while target_mag >= min_mag and prev_mag > Mag(1):
+            assert prev_mag > target_mag
+            assert target_mag.to_layer_name() not in self.mags
+
+            prev_mag_ds = self.mags[prev_mag.to_layer_name()]
+
+            mag_factors = [
+                t / s for (t, s) in zip(target_mag.to_array(), prev_mag.to_array())
+            ]
+
+            # initialize the new mag
+            target_mag_ds = self._initialize_mag_from_other_mag(
+                target_mag, prev_mag_ds, compress
+            )
+
+            # Get target view
+            target_mag_view = target_mag_ds.get_view(is_bounded=not compress)
+
+            # perform upsampling
+            with get_executor_for_args(args) as executor:
+                voxel_count_per_cube = np.prod(prev_mag_ds._get_file_dimensions())
+                job_count_per_log = math.ceil(
+                    1024 ** 3 / voxel_count_per_cube
+                )  # log every gigavoxel of processed data
+
+                if buffer_edge_len is None:
+                    buffer_edge_len = determine_buffer_edge_len(
+                        prev_mag_ds.view
+                    )  # DEFAULT_EDGE_LEN
+                func = named_partial(
+                    downsample_cube_job,
```

Comment:
The name of the method is not good. It was previously just used for downsampling but I reused it for upsampling. Do you have a better idea than `up_or_downsample_cube_job`?

Reply:
I think `up_or_downsample_cube_job` would be fine. However, I am not sure if it would be better to just make 2 functions, which would get rid of the `upsample` flag and some if/else branches.

Reply:
@philippotto do you have a take on that?

Reply:
regarding the name, "resample_cube_job" comes to mind. didn't look into whether it would be better to have two distinct functions, though.

Reply:
Or was the question directed towards the latter? I think @rschwanhold can decide this better, since he worked with the code.

Reply:
Ok. fine with me.